internal/scheduler/legacyscheduler.go

```diff
@@ -0,0 +1,399 @@
+	// Minimum quantity allowed for jobs leased to this cluster.
+	MinimumJobSize map[string]resource.Quantity
+	// These factors influence the fraction of resources assigned to each queue.
+	PriorityFactorByQueue map[string]float64
+	// Base random seed used for the the scheduling loop.
+	// The first iteration uses InitialSeed + 1, the second InitialSeed + 2, and so on.
+	InitialSeed int64
+	// Store reports for each scheduling attempt.
+	JobSchedulingReportsByQueue map[string]map[uuid.UUID]*JobSchedulingReport
+}
+
+// Schedule is similar to distributeRemainder, but is built on NodeDb.
+func (c *LegacyScheduler) Schedule(
+	ctx context.Context,
+	initialUsageByQueue map[string]schedulerobjects.QuantityByPriorityAndResourceType,
+) ([]*api.Job, error) {
+	log := ctxlogrus.Extract(ctx)
+
+	// Total resource usage across all priorities by queue.
+	totalResourcesByQueue := make(map[string]schedulerobjects.ResourceList)
+	for queue, quantityByPriorityAndResourceType := range initialUsageByQueue {
+		totalResourcesByQueue[queue] = quantityByPriorityAndResourceType.AggregateByResource()
+	}
+
+	// Initialise reports dict if not done already.
+	if c.JobSchedulingReportsByQueue == nil {
+		c.JobSchedulingReportsByQueue = make(map[string]map[uuid.UUID]*JobSchedulingReport)
+	}
+
+	// Total resources assigned during this invocation of the scheduler.
+	roundResources := schedulerobjects.ResourceList{
+		Resources: make(map[string]resource.Quantity),
+	}
+
+	// Total resources assigned to each queue during this invocation of the scheduler.
+	roundResourcesByQueue := make(map[string]schedulerobjects.ResourceList)
+
+	// Jobs to lease for each queue.
+	leasedJobsByQueue := make(map[string][]*api.Job)
+
+	// Track the total number of jobs to lease.
+	numJobsToLease := 0
+
+	// To reduce the number of calls to Redis,
+	// we retrieve jobs in batch and store jobs in-memory.
+	jobCacheByQueue := make(map[string][]*api.Job)
+
+	// Used to return early if the scheduler makes no progress.
+	consecutiveIterationsWithNoJobsLeased := 0
+
+	// Maps queue name to a bool, which is true if there are no more
+	// jobs in this queue, beyond those already downloaded.
+	gotAllQueuedJobsByQueue := make(map[string]bool)
+
+	// Random seed used when selecting queues.
+	// Provided as a paremeter to ensure the scheduler is deterministic when testing.
+	seed := c.InitialSeed
+
+	// Schedule jobs one at a time.
+	for (c.SchedulingConfig.MaximumJobsToSchedule == 0 || numJobsToLease < c.SchedulingConfig.MaximumJobsToSchedule) && len(totalResourcesByQueue) > 0 && consecutiveIterationsWithNoJobsLeased < len(totalResourcesByQueue) {
+
+		// Return early if the context deadline has expired.
+		select {
+		case <-ctx.Done():
+			break
+		default:
+		}
+
+		fmt.Println()
+		fmt.Println("=======")
+
+		fmt.Println("iteration ", seed)
+
+		// Select a queue to schedule job from.
+		// Queues with fewer resources allocated to them are selected with higher propability.
+		seed += 1
```

Comment:
I'm not sure this is going to give as much randomness as we want.  I think it would be much better if we just made a threadsafe `rand.Rand()` and injected that.